### DISCUSSION TITLE
Exporting enumerations

### DISCUSSION THREAD

**Discussion by migueldeicaza** [migueldeicaza/SwiftGodot#473] [General]
Hello,

One of the tasks that has been on the back burner for a while is to expose variables that are enumerations to Godot, this issue has been tracked in https://github.com/migueldeicaza/SwiftGodot/issues/128 and I have now a proof of concept and would love some testing.

The code lives in the `exportEnum` branch.

To use it, make your enumeration `CaseIterable` and also make it conform to `Int`:

```
enum MyEnumeration: Int, CaseIterable {
    case first
    case second
}
```

And then to export this, use the `.enum` parameter to Export:

```
@Godot
class Demo: Node {
     @Export(.enum)
     var myState: MyEnumeration
}
```

Would love for people to take this for a spin, and if it works, I can merge it into main.

*Update:* Originally, I had thought about using `ExportEnum`, and Brandon Munshaw suggested using `Export(.enum)` instead, so I went with that option.

**Comment by munshaw** [comment]
The following works beautifully:
```
enum Suite: Int, CaseIterable {
    case spades, hearts, clubs, diamonds
}

enum Rank: Int, CaseIterable {
    case two, three, four, five, six, seven, eight, nine, ten, jack, queen, king, ace
}

@Godot
class Card: Node {
    @Export(.enum) var suite: Suite = .spades
    @Export(.enum) var rank: Rank = .ace
}
```

However, adding `@Export(.enum) var corner: Corner = .bottomLeft` results in the attached error.

[error.txt](https://github.com/migueldeicaza/SwiftGodot/files/15381093/error.txt)

**Comment by migueldeicaza** [comment]
Thanks for the feedback.

I have updated the branch with additional fixes.  One of the limitations is that exporter enumerations must be backed by an Int64.

**Comment by migueldeicaza** [comment]
The version that got merged no longer requires the enumeration to be backed by an Int64, it will now cast to any integral value.
